src/JuliusSweetland.OptiKey.Core/UI/ViewModels/MainViewModel.ServiceEventHandlers.cs

```diff
@@ -2372,46 +2372,152 @@ private void SelectVoice(string voice)
             NavigateToMenu();
         }
 
-        private void StepList(string command)
+        //public static void Main()
+        public void TimerStepList(Commands cmd)
+        {
+            //send the looping key commands to the TimerList
+            TimerList.Add(cmd);
+            System.Timers.Timer aTimer = new System.Timers.Timer();
+            aTimer.Elapsed += new ElapsedEventHandler(OnTimedEvent);
+            aTimer.Interval = 100;
+            aTimer.Enabled = true;
+            //aTimer.Dispose(); // Cancel the timer now
+        }
+        // Specify what you want to happen when the Elapsed event is raised.
+        private void OnTimedEvent(object source, ElapsedEventArgs e)
+        {
+            //check for any new commands
+            if (TimerList.Count() > 0)
+            {
+                //handle the commands
+                StepListCommands(TimerList.First().SingleKeyValue, TimerList.First().MultiKeySelection, TimerList.First().CommandList);
+                //remove the commands
+                TimerList.RemoveAt(0);
+            }            
+        }
+        public class Commands
+        {
+            public KeyValue SingleKeyValue { get; set; }
+            public List<string> MultiKeySelection { get; set; }
+            public List<string> CommandList { get; set; }
+        }
+        public List<Commands> TimerList = new List<Commands>();
+
+        private void StepList(KeyValue singleKeyValue, List<string> multiKeySelection)
         {
-            if (!string.IsNullOrEmpty(command))
+            if (!string.IsNullOrEmpty(singleKeyValue.String))
             {
-                KeyValue singleKeyValue;
-                string[] stringSeparators = new string[] { "<" };
-                foreach (var vStep in command.Split(stringSeparators, StringSplitOptions.None).ToList())
+                Log.InfoFormat("StepList called with singleKeyValue [{0}]", singleKeyValue.String);
+
+                List<string> commandList = singleKeyValue.String.Split('<').ToList();
+                if (commandList.Count > 0)
                 {
-                    Log.DebugFormat("Performing StepList step: {0}.", vStep);
-                    if (vStep.StartsWith("Action>") &&
-                        (Enum.TryParse<FunctionKeys>(vStep.Substring(7), out FunctionKeys result)))
+                    //old but works:
```

Comment:
Here is the synchronous processing of a key. If you want a key to press repetitively until released this is where you would get stuck and where we need a new thread.

The 'else' branch is where I go mess with timers and threads.